sympy/matrices/expressions/matexpr.py

```diff
@@ -613,6 +621,17 @@ def doit(self, **kwargs):
     def indices(self):
         return self.args[1:]
 
+    def _eval_subs(self, old, new):
+        # when replacing the symbol, the indices cannot
+        # access out of bounds
+        if old == self.symbol:
+            if type(old) == type(new):
+                for i, j in zip(self.indices, new.shape):
+                    if i.is_Integer and j.is_Integer and i >= j:
+                        raise IndexError('indices access out of range')
```

Comment:
I think it would be better to check this in `MatrixElement.__new__` rather than `_eval_subs`.

Reply:
I tried and failed but I think I got thrown off by thinking the MatrixBase clause was for MatrixSymbol...I will try again to make the change there.

Reply:
OK, this is part of `__new__` now.